File: TrainWise.Web/Models/TrainRequest.cs
namespace TrainWise.Web.Models;

public sealed class TrainRequest
{
    public string DatasetId { get; set; } = string.Empty;
    public string TargetColumn { get; set; } = string.Empty;
    public string TaskType { get; set; } = "classification";
    public PreprocessingConfig Preprocessing { get; set; } = new();
    public ModelConfig Model { get; set; } = new();
    public double TrainTestSplit { get; set; } = 0.8;
    public bool CrossValidation { get; set; }
}

public sealed class PreprocessingConfig
{
    public string NullStrategy { get; set; } = "drop";
    public string Encoding { get; set; } = "onehot";
    public string Scaling { get; set; } = "standard";
    public bool ApplySmote { get; set; }
    public List<string> ExcludeColumns { get; set; } = new();
}

public sealed class ModelConfig
{
    public string Name { get; set; } = "LogisticRegression";
    public ModelHyperparameters Hyperparameters { get; set; } = new();
}

public sealed class ModelHyperparameters
{
    public int? NEstimators { get; set; }
    public double? C { get; set; }
    public string? Kernel { get; set; }
    public int? NNeighbors { get; set; }
    public int? MaxIter { get; set; }
}

File: TrainWise.MLService/app/models/train_request.py
from typing import Literal, List, Optional, Dict, Any
from pydantic import BaseModel, Field

class PreprocessingConfig(BaseModel):
    nullStrategy: Literal["drop", "mean", "median", "mode"]
    encoding: Literal["onehot", "label"]
    scaling: Literal["standard", "minmax", "none"]
    applySmote: bool = False
    excludeColumns: List[str] = Field(default_factory=list)

class ModelHyperparameters(BaseModel):
    n_estimators: Optional[int] = None
    C: Optional[float] = None
    kernel: Optional[str] = None
    n_neighbors: Optional[int] = None
    max_iter: Optional[int] = None
    alpha: Optional[float] = None
    max_depth: Optional[int] = None
    additionalParams: Optional[Dict[str, Any]] = Field(default_factory=dict)

class ModelConfig(BaseModel):
    name: str # Simplified to allow more models without literal restriction
    hyperparameters: ModelHyperparameters = Field(default_factory=ModelHyperparameters)

class TrainRequest(BaseModel):
    datasetId: str
    targetColumn: str
    taskType: Literal["classification", "regression"]
    preprocessing: PreprocessingConfig
    model: ModelConfig
    trainTestSplit: float = 0.8
    crossValidation: bool = False
    kFolds: int = 5
    filePath: Optional[str] = None
